### ISSUE TITLE
Problem using mvn at building time

### ISSUE THREAD

**Issue by doublebyte1** [carlossg/docker-maven#32]
I have the following use case:
1 - In the Dockerfile I  build a mvn project and generate some jars
2 - then I go and build another project which uses the previous jars as dependencies

with the /root/.m2 volume, I find that the dependency jars (build on step 1), are gone, so I cannot successfully build the second project. This is maybe because the Dockerfile has layers, and the repository is not persisted in the second layer?

I already tried the approach that you advise in the README (using the settings.xml, as persisted repository), but I think this approach is designed to persist the repositories *after* the build of the Dockerfile , on runtime. I can see this image is designed to use maven on runtime, but what would be the correct way to use it also during building time?

**Comment by carlossg** [comment]
Nothing in /root/.m2 is persisted in the image, you would need to put the jars somewhere else

**Comment by doublebyte1** [comment]
Thanks for the quick reply. Based on what you said, this is what worked for me (maybe it can help others). 

As the repository defined on settings.xml is only available on runtime, I have set it explicitly during my builds. Building the first project:

`RUN mvn install -DskipTests -Drelax -gs /usr/share/maven/ref/settings-docker.xml `

This has left the jars on `/usr/share/maven/ref/repository`, according to what is defined in settings.xml. Then I built the second project with the same flag:

`RUN mvn install -DskipTests -Drelax -gs /usr/share/maven/ref/settings-docker.xml `

And it found the dependencies on `/usr/share/maven/ref/repository`

**Comment by KevinSilvaQuintana** [comment]
Hi @doublebyte1, @carlossg. I am actually having a similar issue. How did you customize the contents of setting-docker.xml inside of your image? I am trying to set up the build on a CI with the following image (https://github.com/alirizasaral/Maven-with-Proxy) and I have the issue that maven downloads all the dependencies from scratch without caching them inside the repository. Should I not be using an image based on alpine to get the maven repository to work correctly?

**Comment by cedricwalter** [comment]
@KevinSilvaQuintana 

Add in your Dockerfile
`COPY settings-docker.xml /usr/share/maven/ref/`
in your docker directory, create a new file `settings-docker.xml` with content
```
<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
  | xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  | xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
  | https://maven.apache.org/xsd/settings-1.0.0.xsd">
  | <localRepository>/usr/share/maven/ref/repository</localRepository>
  | </settings>
```
use in all mvn command
`RUN mvn install -DskipTests -Drelax -gs /usr/share/maven/ref/settings-docker.xml`

run again a build with no-cache to avoid to reuse some step that were not creating dependency jar in new repository folder
` sudo docker build --no-cache <yourdockerdirectory>`

**Comment by KevinSilvaQuintana** [comment]
Hi @cedricwalter ,

I actually ended up finding a solution which I have outlined in this post:
https://stackoverflow.com/questions/33430487/how-to-use-gitlab-ci-to-build-a-java-maven-project

However, between that and your solution, I am not sure which is more 'clean'. Your approach is very simple and straight forward. However, I think that being able to centralize everything in the CI build file will allow for a more generic re-use for future projects. I'll give it a try though. :+1:
